### ISSUE TITLE
Cells not rendering or loading and code seems lost in jupyterlab notebook

### ISSUE THREAD

**Issue by mariajmolina** [jupyterlab/jupyterlab#10840]
Hello! I frequently use JupyterLab and notebooks within the platform. 

Certain cells that contain code can stop loading at times, and I am unable to click on them or view the code contained in the cells (see screenshot). 

Has anyone else encountered this? Is this related to internet connection or something else? It took about 10 minutes for the cells to show up despite reloading the notebook, stopping and restarting the kernel, etc. 

Thanks for your help!

<img width="741" alt="Screen Shot 2021-08-16 at 1 47 37 AM" src="https://user-images.githubusercontent.com/24659645/129530058-b6197e3a-6b9e-435f-8870-a094232d5c68.png">

**Comment by krassowski** [comment]
Thanks for opening this issue. It sounds like it could be related to a new performance optimization in JupyterLab 3.1 not working as intended. Could you please confirm which exact version of JupyterLab you are using (you can get it from the top menu bar "Help" → "About JupyterLab")?

**Comment by mariajmolina** [comment]
Thanks so much, @krassowska ! The version of JupyterLab I am using is 3.1.6. Please let me know if you need any additional information.

**Comment by krassowski** [comment]
Thank you. A few more questions:

1. When this happens, is there any error message in the [browser developer console](https://webmasters.stackexchange.com/a/77337)? That would be very helpful to determine.
2. How many cells is there in your notebook? Is there a notebook in which it happened repeatedly? If so, would you possibly be able to share such a notebook?
3. What browser do you use? What is its version?
4. Could you please go to `Advanced Settings Editor` -> `Notebook` and in the right pane (User Preferences) paste the following:

  ```
  {
      "renderCellOnIdle": false,
      "numberCellsToRenderDirectly": 10000000000000
  }
  ```
  
  and see if the problem goes away after pressing save and reloading the notebook?

**Comment by mariajmolina** [comment]
Some answers:

1. Yes! There are quite a few error messages, I am not sure if all are related to this. Screenshots below. Screenshots were taken before the notebook was reloaded and issue (at least temporarily) resolved.

2. There are 87 cells. This issue has happened a few times, but I would not say it is consistent. I raised this github issue because I was scared that I lost my code in the respective cells(!), but then they eventually came back (reloaded). I have not uploaded the notebook to github, but can do so tomorrow and link here if needed.

3. I use google chrome, and upon checking the version, the browser updated. It is now Version 92.0.4515.131 (Official Build) (x86_64). Upon updating and relaunching the browser, the issue appears to have gone away in my JupyterLab notebook, so I am not sure if it was related to the previous google chrome version, or if I just needed to relaunch because my browsers/tabs were open for too many days.

4. I have gone ahead and done this -- thanks for proposing a solution! Upon doing this, all of the cells are present and code is visible (and "run-able").

<img width="1279" alt="Screen Shot 2021-08-16 at 12 23 32 PM" src="https://user-images.githubusercontent.com/24659645/129612440-ada7c594-0ecf-4dc9-b2c9-2f40c51e5942.png">

<img width="1276" alt="Screen Shot 2021-08-16 at 12 23 43 PM" src="https://user-images.githubusercontent.com/24659645/129612453-a8b3ecbf-e90d-407f-893d-f352fcef3703.png">

**Comment by krassowski** [comment]
Thank you very much! The notebook might help if you can provide it (after removing any content you would prefer not to share - I think that we would be just interested in the layout of cells i.e. how many consecutive code/markdown cells there are and which cells were affected). CC @echarles who may have more ideas.

**Comment by krassowski** [comment]
I further narrowed it down in https://github.com/jupyterlab/jupyterlab/issues/10853#issuecomment-899853917 and I _think_ it is related to naive casting of cell to code cell (but we might be getting a placeholder instead). I can imagine that adding a cell type check could alleviate the issue for now, but I am not sure if that would be a good general solution, or if there is anything else we need to do. Still, a good reproducer would help a lot.

**Comment by echarles** [comment]
> I can imagine that adding a cell type check could alleviate the issue for now, but I am not sure if that would be a good general solution, or if there is anything else we need to do. Still, a good reproducer would help a lot.

Hi, I have never seen the error printed in the console. Looks like it is related to SVG rendering. A Cell type check would avoid the stacktrace, but I don't think it will help to get the cell rendered.

> I have not uploaded the notebook to github, but can do so tomorrow and link here if needed.

As @krassowski mentioned, a reproducer will help to understand the issue. @mariajmolina Looking forward your notebook so we can work on it. Thx a lot.

**Comment by TomAugspurger** [comment]
I believe that https://github.com/microsoft/PlanetaryComputerExamples/blob/main/tutorials/landcover.ipynb reproduces the issue (apologies, it's a bit large and requires a GPU to execute... If I spot another example I'll add it).

The issues start showing up around "Predicting land cover for each pixel"

![image](https://user-images.githubusercontent.com/1312546/129804466-4f3ba29c-8722-4af7-8ce8-b173722df8d7.png)

**Comment by zaneselvans** [comment]
I think I might be having the same issue. When I do "run all cells" in notebooks that have previously been fine, there are now a bunch of cells that get skipped and don't run. If I step through one cell at a time, the cells will run, but then the cell focus becomes ambiguous and if I hit shift-enter again, the previous output is lost. Sometimes whole cells vanish.

Currently:
* I'm running JupyterLab 3.1.7
* I have disabled all of my extensions
* I'm running the notebook in a private browsing window on Firefox.
* The problem doesn't show up if I run the same notebook in "classic" mode (not jupyterlab)

I've seen the issue in several notebooks, but haven't been able to create a new one that's standalone and replicates the problem. However, it does show up in [this notebook](https://github.com/catalyst-cooperative/pudl-examples/blob/main/notebooks/03-pudl-parquet.ipynb) starting with code cell 11 which follows the heading "Simple Filtering" and contains:

```py
slow_epacems_dd = dd.read_parquet(epacems_path)
slow_epacems_dd = slow_epacems_dd[
    (slow_epacems_dd.year == 2019)
    & (slow_epacems_dd.state == "ID")
]
```

The next cell that it attempts to run follows the header "Fuel Consumption" which contains:
```py
unit_ids = comanche_df.unitid.unique()
fig, axs = plt.subplots(nrows=len(unit_ids), ncols=1, sharex=True, figsize=(20,15))
for n, unitid in enumerate(comanche_df.unitid.unique()):
    axs[n].scatter(
        comanche_df.query("unitid==@unitid").operating_datetime_utc,
        comanche_df.query("unitid==@unitid").heat_content_mmbtu,
        s=0.2, alpha=0.1,
    )
    axs[n].set_ylim(-100,10_000)
    axs[n].set_title(f"Comanche Unit {unitid}")
    axs[n].set_ylabel("Fuel consumed [mmbtu / hr]")
plt.show();
```

However, if after the first run of the notebook in which cells are skipped I then restart-kernel-and-run-all-cells, the notebook executes normally, running every cell.

**Comment by echarles** [comment]
@mariajmolina @TomAugspurger  @zaneselvans  Thx a lot for reporting. Can you confirm if you face the issue on the initial page load, or when you invoke "Run All Cells".

I have tried to linked notebook, but the need dependencies that don't install on my env. I the simple reproducer is not possible, the error shown in the browser console will help.

**Comment by echarles** [comment]
Regarding my previous comment, I just wanted to make sure the issue was arising on the initial notebook load (it should be, I have been confused reading comment).

> If I downgrade to JupyterLab v3.1.4 the problem no longer seems to happen.

This is interesting. @mariajmolina @TomAugspurger @zaneselvans could I ask you to downgrade step by step from 3.1.7 to 3.1.0 and report when the issue does not arise anymore. Also the full stacktrace in the browser console will help.

**Comment by echarles** [comment]
I have downloaded the following notebooks and opened them in Chrome, Firefox and Safari (on MacOS). They render correctly, without cells remaining unrendered. I can not run them as I am missing python dependencies on my environment, but the virtual notebook rendering is only used on the page load, not on cell run.

- https://github.com/catalyst-cooperative/pudl-examples/blob/main/notebooks/03-pudl-parquet.ipynb
- https://github.com/microsoft/PlanetaryComputerExamples/blob/main/tutorials/landcover.ipynb

Maybe the indication of a specific browser or operating system can help to identify the cause of the issue?

**Comment by mariajmolina** [comment]
Hi @echarles -- Apologies for the delay in getting back to you -- I have not been able to reproduce this issue in any of my JupyterLab notebooks today. The issue went away after my Chrome (on MacOS) updated (my updates are scheduled via a work scheduler) and I know that I was overdue for the update when this JupyterLab issue was occurring. Currently, I am using Chrome Version 92.0.4515.131 (Official Build) (x86_64). I use MacOS Catalina version 10.15.7. There is now another Chrome update scheduled that I am delaying temporarily. The issue was not occurring for me on the initial page load (JupyterLab notebook), but rather after working for a while with the respective notebook. It don't think was specific to any one notebook. The issue would pop up if I scrolled away from a few cells to work on other cells, and then returned to the earlier cells (the earlier cells would show up as those lines in the screen shot I shared when I opened this issue). Reloading the Chrome page wouldn't resolve the issue. Sometimes it resolved on its own after a while, which was also odd. I am working on JupyterLab all week so I will post an update if this happens again, but it appears to have been resolved for me since the Chrome update.

**Comment by mariajmolina** [comment]
I should also add that I am using JupyterLab Version 3.0.16.

**Comment by krassowski** [comment]
Hi @mariajmolina just to confirm, are you running with the settings I suggested in point four of https://github.com/jupyterlab/jupyterlab/issues/10840#issuecomment-899661421 enabled? If yes, then you will not be able to reproduce the issue anymore (unless having a really huge notebook) as effectively turn off the mechanism which is contributing to the problem. If you want to test a redacted version of a notebook, first remove these settings temporarily and reload JupyterLab.

**Comment by TomAugspurger** [comment]
I might have messed something up, but I was only able to reproduce the cell rendering bug when jupyterlab via a (remote) JupyterHub session.


```python
$ git clone https://github.com/microsoft/PlanetaryComputerExamples
$ cd PlanetaryComputerExamples
$ docker run -it --rm \
    -v $PWD:/home/jovyan \
    -p 8888:8888 \
    mcr.microsoft.com/planetary-computer/python:2021.08.16.0 \
    jupyter lab --no-browser --ip="0.0.0.0"
```

That's the image running on my Hub's singleuser, so I expected that opening the `tutorials/landcover.ipynb` notebook would reproduce the error locally (in that docker container), but it didn't.

Like I said though, I could have easily messed something up when testing it, so I wouldn't put too much faith in JupyterHub / remote kernels mattering.

**Comment by zaneselvans** [comment]
@echarles I was only seeing the weird rendering behavior when I attempted to run the cells. Opening the notebook up it always looked fine, until getting to one of the problematic cells and trying to run it, at which point the input/output number gets blanked out, and that weird horizontal line appears between cells. It would also not move the focus to the next cell -- no cell would be focused.

Stepping through one cell at a time and just hitting shift-enter repeatedly, I ended up with only every-other cell visible (just the odd numbered inputs/outputs).

**Comment by ajye92** [comment]
I have encountered the same issue for a couple of small ipynb files with Jupyterlab version 3.1.6 and 3.1.7. There was no error messages in the console window. One additional issue I had was that "Run All Cells" may skip a few cells in the middle or at the end of the file. I can confirm that version 3.1.4 has no such issues. I tried Safari, Chrome, and Edge in MacOS Big Sur but with no luck.

**Comment by echarles** [comment]
Thx all for the reports. It sounds like the issue is not there on 3.1.4 but for 3.1.6 and 3.1.7.

@krassowski FWIW 3.1.4 had already the virtual notebook merged in https://github.com/jupyterlab/jupyterlab/blob/v3.1.4/packages/notebook/src/widget.ts#L198-L219

The only change after that for the virtual notebook in 3.1.x branch is https://github.com/jupyterlab/jupyterlab/pull/10797. I am trying to make sense how this change could generate the reported issue. So far, I can not make sense and I am a bit block on that.

I don't see an easy path forward without a reproducer.

**Comment by krassowski** [comment]
Here is one reproducible issue (I doubt it is a root cause but it looks related to some of the reports in here):

**When notebook is not full rendered then cells do not get rendered to proper state after "Restart and run all cells"**

To reproduce reliably:
1. use the binder image for convenience [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gist/krassowski/fd91a4124fd204125545c62ab89683e4/HEAD?urlpath=lab); it is based on this [gist](https://gist.github.com/krassowski/fd91a4124fd204125545c62ab89683e4#file-50_cells-ipynb) you can fork it and adjust it to your liking
2. set notebook settings to https://gist.github.com/krassowski/fd91a4124fd204125545c62ab89683e4#file-settings-json to minimise unpredictable behaviour
3. set browser viewport to a height such that no more than 10 cells of https://gist.github.com/krassowski/fd91a4124fd204125545c62ab89683e4#file-50_cells-ipynb notebook are shown
4. reload JupyterLab
5. open the [50_cells-ipynb](https://gist.github.com/krassowski/fd91a4124fd204125545c62ab89683e4#file-50_cells-ipynb) notebook
6. open search bar to confirm that the notebook is not fully rendered ("This document is still loading. Only loaded content will appear in search results until the entire document loads.")
7. press the "restart and re-run" button (fast-forward icon)
8. wait until execution is complete
9. scroll down to see that cells are not properly rendered (the rendering shows cells prior to being run)
10. run any cell once again to confirm that all cells were in fact run and this is a rendering issue; the execution count will show `51`.

GIF:

![reproducer-rendering-restart-and-run-all](https://user-images.githubusercontent.com/5832902/130325912-5f259c7f-6ad1-4906-831d-4e65bcae946d.gif)

That one is reproducible on both 3.1.7 and 3.1.0. I assume all versions in between are affected as well.

>  It sounds like the issue is not there on 3.1.4 but for 3.1.6 and 3.1.7.

Interesting... I am not convinced - lack of reproduction on 3.1.4 does not mean it is not there ;) The description of this issue matches one of the issues I was experiencing already in 3.1 beta so my instinct says the issue might have been already there in 3.1.0 (but maybe 3.1.4 or something around made it trigger more often; and obviously I can be wrong).

**Comment by krassowski** [comment]
And there is some traceback in the console:

```js
jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2 Uncaught (in promise) TypeError: Cannot set property 'future' of undefined
    at Function.t [as execute] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at Function.t [as execute] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at m (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2
    at Array.map (<anonymous>)
    at Object.u [as runSelected] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at Function._ [as runAll] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2
```

and

```js
jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2 Uncaught (in promise) TypeError: Cannot read property 'future' of undefined
    at Function.t [as execute] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at m (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2
    at Array.map (<anonymous>)
    at Object.u [as runSelected] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at Function._ [as runAll] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2
```

**Comment by krassowski** [comment]
A very similar effect can be achieved when running all cells by pressing Shift + Enter and keeping it pressed for a dozen of seconds:

```js
jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2 Uncaught (in promise) TypeError: Cannot set property 'future' of undefined
    at Function.t [as execute] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at Function.t [as execute] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at m (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2
    at Array.map (<anonymous>)
    at Object.u [as runSelected] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at Function.C [as runAndAdvance] (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at run (jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2)
    at jlab_core.e9e49407afea635b7583.js?v=e9e49407afea635b7583:2
```

**Comment by krassowski** [comment]
Some thoughts:
- should there be a semaphore for `_renderPlaceholderCell`? Looking at the code it seems that if both the intersection observer and idle renderer (`renderPlaceholderCells`) kick in at the same time, this could potentially lead to out-of-order rendering.
- should we increase the default `numberCellsToRenderDirectly` to say 100 (we talked about it on Wednesday as a possible mitigation strategy for infancy issues of the renderer)?

**Comment by echarles** [comment]
> should there be a semaphore for _renderPlaceholderCell? Looking at the code it seems that if both the intersection observer and idle renderer (renderPlaceholderCells) kick in at the same time, this could potentially lead to out-of-order rendering.

Without testing, but just from what I have in mind of the current implementation, the case you describe, "Restart and Run All Cells" when all the cells are not yet fully rendered can certainly cause issues like the one described.

In terms of solution, the semaphore may be a good step, but I am not sure it would be enough to avoid all side effects (need to test). If this is not enough, another solution would be to disallow restart while the cells are still rendering (I doubt this option will get fans...).

> should we increase the default numberCellsToRenderDirectly to say 100 (we talked about it on Wednesday as a possible mitigation strategy for infancy issues of the renderer)?

That can be done.

**Comment by krassowski** [comment]
@ajye92, @zaneselvans, @Doggie52 and anyone else experiencing this issue, if you would like to help to narrow down this issue you can dissect it by testing the following online instances of JupyterLab (on Binder) which correspond to commits made between 3.1.4 and 3.1.6 from the oldest (closest to 3.1.4) to the newest:

1. ["Markdown url resolver no longer throws for malformed URLs in isLocal check"](https://mybinder.org/v2/gh/meeseeksmachine/jupyterlab/auto-backport-of-pr-10773-on-3.1.x?urlpath=lab-dev)
2. ["Revert input guard"](https://mybinder.org/v2/gh/meeseeksmachine/jupyterlab/auto-backport-of-pr-10779-on-master?urlpath=lab-dev) 3.1.5 was meant to be published after this was one but was skipped
3. ["Support collapsible headers with virtual notebook rendering"](https://mybinder.org/v2/gh/meeseeksmachine/jupyterlab/auto-backport-of-pr-10793-on-3.1.x?urlpath=lab-dev)
4. ["Share nbformat and nbformatMinor"](https://mybinder.org/v2/gh/meeseeksmachine/jupyterlab/auto-backport-of-pr-10795-on-3.1.x?urlpath=lab-dev)
5. ["Add translations for kernel statuses"](https://mybinder.org/v2/gh/krassowski/jupyterlab/auto-backport-of-pr-10738-on-3.1.x?urlpath=lab-dev) 3.1.6 was published immediately after this one

**Comment by krassowski** [comment]
I think this is now narrowed down the the link number 3, please see https://github.com/jupyterlab/jupyterlab/issues/10895#issuecomment-903960185.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/crowdsourcing-for-a-blog-cool-binder-features/10462/3

**Comment by krassowski** [comment]
Hi all, please give the recent 3.1.10 a try, a fix by @echarles was included, and it would be great to have confirmation on the state of the matters.

**Comment by ajye92** [comment]
Thanks! It appears to me that the problem has been solved in 3.1.10.

**Comment by krassowski** [comment]
Thank you for confirmation that JupyterLab 3.1.10 fixed the most prevalent issue!
I will close this one to focus discussion on remaining edge cases in https://github.com/jupyterlab/jupyterlab/issues/10933 (please feel welcome to leave your feedback there).

**Comment by BlaneG** [comment]
I don't know how I ran into this problem, but here was my work around.

In the JupyterLab interface horizontal lines appear in place of where input and output cells should be appearing. When inspecting the .ipynb with a text editor, I noticed that the cells that were appearing as horizontal lines in JupyterLab interface were code cells that contained the metadata: 
```
   "execution_count": null,
   "outputs": []
```

I could open the .ipynb with VSCode and the cells that were not visible in the JupyterLab interface were again visible within vscode, but these cells contained no output. If I executed the notebook in vscode, all of the cell outputs were regenerated including the cells that were appearing as horizontal lines in the JupyerLab interface. When I save this result and open the notebook with the JupyterLab interface, all of the cells are visible again, but I receive the "Code Editor Out of Sync" error message which can be resolved by replacing `\r\n\` with `\n` in the notebook file as discussed [here](https://stackoverflow.com/questions/65673839/how-can-i-fix-a-jupyterlab-code-editor-out-of-sync-error-message).
